### PR TITLE
HPCC-21062 Add ECL syntax for DISTRIBUTE(ds, ALL)

### PR DESCRIPTION
**Pull request by ghalliday**
Also implement much of the underlying code to support

    DISTRIBUTE(ds, f(LEFT, TARGETNODE))

to allow distribution to a set of nodes

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [x] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by ghalliday** [comment]
I'm not sure Set Distribute is the best name.  I have used NWay Distribute within the code generator.  What do we want the activity to be called?

@jakesmith here is the code generator work - check it does what you need
@shamser please review.

**Comment by jakesmith** [comment]
>I'm not sure Set Distribute is the best name. I have used NWay Distribute within the code generator. What do we want the activity to be called?

I think NWay Distribute is an improvement.
"distribute" is a bit misleading, it is more of a copy, clone or propagate.

**Review by shamser — APPROVED** [review]
@ghalliday Code Reiewed.  It would be useful to have the documentation Jira linked in this ticket. As far as  whether it's called DISTRIBUTE or NWAYDISTRIBUTE - not really fussed.

**Comment by ghalliday** [comment]
@richardkchapman renamed to nway distribute and repushed.

File: ecl/hqlcpp/hqlhtcpp.cpp [hunk 1]
```diff
@@ -6146,6 +6146,7 @@ double HqlCppTranslator::getComplexity(IHqlExpression * expr, ClusterType cluste
             complexity = 5;
         break;
     case no_distribute:
+    case no_nwaydistribute:
```
Comment (richardkchapman) [inline ecl/hqlcpp/hqlhtcpp.cpp]: Not sure what the complexity implies, but is it really appropriate to be the same as a normal distribute?

**Comment by richardkchapman** [comment]
@ghalliday one trivial question

File: ecl/hqlcpp/hqlhtcpp.cpp (hunk 1 above)
Comment (ghalliday) [reply ecl/hqlcpp/hqlhtcpp.cpp]: This code is used to calculate the "complexity" of a query.  I'm not convinced it is very useful, but treating it the same as a distribute is a reasonable assumption.

**Comment by ghalliday** [comment]
@richardkchapman replied

**Pull request merged**